torch/csrc/jit/codegen/cuda/scheduler/pointwise_utils.cpp

```diff
@@ -0,0 +1,99 @@
+#include <torch/csrc/jit/codegen/cuda/scheduler/pointwise_utils.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+namespace pointwise_utils {
+
+DomainMap::DomainMap(Fusion* fusion)
+    : fusion_(fusion), ca_map_(ComputeAtMap(fusion)) {
+  view_tvs_ = scheduler_utils::getViewTVs(fusion);
+}
+
+bool DomainMap::areMapped(IterDomain* id1, IterDomain* id2) {
+  return ca_map_.areMapped(id1, id2, IdMappingMode::EXACT);
+}
+
+// Determine if all iterDomains are mapped between input and output tvs
+bool DomainMap::areAllMapped(TensorView* input_tv, TensorView* output_tv)
+    const {
+  // Get concrete IDs for input root or rfactor domain
+  std::unordered_set<IterDomain*> in_concrete_ids;
+  for (auto in_id : input_tv->getMaybeRFactorDomain()) {
+    if (!ca_map_.getConcreteMappedID(in_id, IdMappingMode::EXACT)
+             ->isBroadcast() &&
+        !in_id->isReduction()) {
+      in_concrete_ids.insert(
+          ca_map_.getConcreteMappedID(in_id, IdMappingMode::EXACT));
+    }
+  }
+
+  // Erase all input concrete IDs mapped to the output domain
+  // Ignore unresolved broadcast dimensions
+  for (auto out_id : output_tv->getMaybeRFactorDomain()) {
+    if (!out_id->isBroadcast()) {
+      if (!eraseIfMapped(in_concrete_ids, out_id)) {
+        eraseIfMappedThroughView(in_concrete_ids, out_id);
+      }
+    }
+  }
+  return in_concrete_ids.empty();
+}
+
+// Erase input concrete ID if it is mapped to output ID
+bool DomainMap::eraseIfMapped(
+    std::unordered_set<IterDomain*>& in_concrete_ids,
+    IterDomain* out_id) const {
+  auto out_concrete_id =
+      ca_map_.getConcreteMappedID(out_id, IdMappingMode::EXACT);
+  auto in_concrete_id_iter = in_concrete_ids.find(out_concrete_id);
+  bool found_match = in_concrete_id_iter != in_concrete_ids.end();
+  if (found_match) {
+    in_concrete_ids.erase(in_concrete_id_iter);
+  }
+  return found_match;
+}
+
+// Check if in_id is mapped to out_id through any view rfactor domain
+void DomainMap::eraseIfMappedThroughView(
```

Comment:
This function looks weird to me as well:

On the one hand, I don't think it handles the case where there are multiple views. For example, if I have: `x->view->view->y`, then this function could not correctly detect IDs in `x` that are mapped through view to `y`.

On the other hand, I am not sure what "mapped through view" should be defined. For example, if I have
```
T0[I1{2}, I2{2}]
T1[I3{2*2}] = view(T0)
T2[I4{2}, I5{2}] = view(T1)
```
Should I expect `I4` to be mapped to `I1` in the compute-at map? My opinion is "yes".

Also, I have a feeling that we need some transformation simplification mechanism to be able to reason about which ID is mapped to which. In the above example, if I start from `[I1{2}, I2{2}]`, do a merge `[I3{2*2}]`, then a split 2, then our system should be able to identify this and cancel the merge and split. We are not doing so partly because `ceilDiv` is not invertible, but for the case of view, I think what we should use is not `ceilDiv`, but something like below:
```cuda
int exactDiv(int x, int y) {
  assert(x % y == 0);
  return x / y;
}
```
And our system should be able to do the simplification like
```
exactDiv(i0, 2) * 2 --> i0
exactDiv(i0, 4) * 2 --> exactDiv(i0, 2)
exactDiv(i0, 2) * 4 --> i0 * 2
exactDiv(i0 * 4, 2) --> i0 * 2
exactDiv(i0 * 2, 4) --> exactDiv(i0, 2)
```
And I think this type of simplification should also be useful for `view(dtype)` and `view_as_real`.

Reply:
This is an example that the pointwise scheduler could not schedule because it could not find reference.
```c++
TEST_F(NVFuserTest, TMP1) {
  std::unique_ptr<Fusion> fusion_ptr = std::make_unique<Fusion>();
  Fusion& fusion = *fusion_ptr.get();
  FusionGuard fg(&fusion);

  auto tv0 = makeConcreteTensor({2, 3, 5});
  fusion.addInput(tv0);
  auto tv1 = view(tv0, {2, 3, 5}, {6, 5});
  auto tv2 = sin(tv1);
  auto tv3 = view(tv2, {6, 5}, {2, 3, 5});
  fusion.addOutput(tv3);

  ComputeAtMap ca_map(&fusion);
  // fusion.print();
  // std::cout << ca_map.toString() << std::endl;

  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
  at::Tensor input = at::randn({2, 3, 5}, options);

  FusionExecutorCache fec(std::move(fusion_ptr));
  auto cg_outputs = fec.runFusionWithInputs({input});

  testValidate(
      &fusion,
      cg_outputs,
      {input},
      {input.sin()},
      __LINE__,
      __FILE__);
}
```

```
***Runtime***: Try to schedule fusion un-segmented:

Scheduler _reduction_ ***rejected*** because : No support for view op
Scheduler _pointwise_ ***rejected*** because : cannot find reference tensor
Scheduler _persistent_ ***rejected*** because : needs a reduction op
```

So we are only supporting the case where there is an output tensor that has at most one view on its path to input tensors?

Reply:
View is rejected now